pkg/controller/privatelink/actuator/awsactuator/awshubactuator_test.go

```diff
@@ -0,0 +1,1435 @@
+
+		AWSClientConfig func(*mock.MockClient)
+
+		expectError string
+	}{{ // There should be an error on cleanupHostedZone failure
+		name: "failure on cleanupHostedZone",
+		cd: cdBuilder.Build(
+			testcd.WithAWSPlatformStatus(&hivev1aws.PlatformStatus{
+				PrivateLink: &hivev1aws.PrivateLinkAccessStatus{
+					HostedZoneID: testHostedZone,
+				},
+			}),
+		),
+		AWSClientConfig: func(m *mock.MockClient) {
+			m.EXPECT().ListResourceRecordSets(gomock.Any()).Return(nil, awserr.New("AccessDenied", "not authorized to ListResourceRecordSets", nil))
+		},
+		expectError: "error cleaning up Hosted Zone: failed to list the hosted zone testhostedzone: AccessDenied: not authorized to ListResourceRecordSets",
+	}, { // There should not be an error on success
+		name: "success",
+		cd: cdBuilder.Build(
+			testcd.WithAWSPlatformStatus(&hivev1aws.PlatformStatus{
+				PrivateLink: &hivev1aws.PrivateLinkAccessStatus{
+					HostedZoneID: testHostedZone,
+				},
+			}),
+		),
+		AWSClientConfig: func(m *mock.MockClient) {
+			m.EXPECT().ListResourceRecordSets(gomock.Any()).Return(nil, awserr.New("NoSuchHostedZone", "hosted zone not found", nil))
+		},
+	}}
+
+	for _, test := range cases {
+		t.Run(test.name, func(t *testing.T) {
+			logger, _ := testlogger.NewLoggerWithHook()
+			awsHubActuator, err := newTestAWSHubActuator(t,
+				&hivev1.AWSPrivateLinkConfig{},
+				test.cd,
+				[]runtime.Object{},
+				test.AWSClientConfig,
+				logger,
+			)
+			require.NoError(t, err)
+
+			err = awsHubActuator.Cleanup(test.cd, test.cd.Spec.ClusterMetadata, logger)
+
+			if test.expectError == "" {
+				assert.NoError(t, err)
+			} else {
+				assert.EqualError(t, err, test.expectError)
+			}
+		})
+	}
+}
+
+func Test_CleanupRequired(t *testing.T) {
+	awsHubActuator := &AWSHubActuator{}
+	cdBuilder := testcd.FullBuilder(testNameSpace, testClusterDeployment, scheme.GetScheme())
+
+	cases := []struct {
+		name string
+		cd   *hivev1.ClusterDeployment
+
+		expectResult bool
+	}{{ // Cleanup is not required when status.platform is nil
+		name: "status.platform is nil",
+		cd:   cdBuilder.Build(),
+	}, { // Cleanup is not required when status.platform.aws is nil
+		name: "status.platform.aws is nil",
+		cd:   cdBuilder.Options(testcd.WithEmptyPlatformStatus()).Build(),
+	}, { // Cleanup is not required when status.platform.aws.privatelink is nil
+		name: "status.platform.aws.privatelink is nil",
+		cd:   cdBuilder.Options(testcd.WithAWSPlatformStatus(&hivev1aws.PlatformStatus{})).Build(),
+	}, { // Cleanup is not required when deleting a cluster with preserve-on-delete
+		name: "preserve on delete",
+		cd: cdBuilder.GenericOptions(
+			generic.Deleted(),
+		).Options(
+			testcd.WithAWSPlatformStatus(&hivev1aws.PlatformStatus{
+				PrivateLink: &hivev1aws.PrivateLinkAccessStatus{
+					HostedZoneID: testHostedZone,
+				},
+			}),
+			testcd.PreserveOnDelete(),
+			testcd.WithAWSPlatform(&hivev1aws.Platform{
+				PrivateLink: &hivev1aws.PrivateLinkAccess{Enabled: true},
```

Comment:
The logic here is that, a cluster that is configured for privatelink + preserve-on-delete could a) have privatelink enabled set to false, and then b) deleted. Logically, we would want to then clean up the privatelink before we delete the cluster even with preserve-on-delete enabled. This would be the path to take if you wanted to preserve the cluster without preserving the privatelink components.

I would actually argue, we probably shouldn't be preserving the hive-specific privatelink components at all when a cluster is deleted. The cluster doesn't need it, only hive does... and that could be recreated if/when a cluster is ever (re)adopted.

Reply:
> a) have privatelink enabled set to false, and then b) deleted

The actor doing this :arrow_up: would be racing with the controller, which would be cleaning up (un-privatelinking? private-unlinking?) at the behest of a) before b) sets `deletionTimestamp`. If the intent is to preserve the infra without preserving privatelink resources, then a well-behaved version of that actor would wait for hive to confirm cleanup by setting privatelink status to "off" before proceeding with deletion.

But I take your point, and I think I agree with it (given that we're already talking about a pretty unlikely corner case). Also seems like maybe we already discussed this...

> I would actually argue, we probably shouldn't be preserving the hive-specific privatelink components at all when a cluster is deleted.

Now this is interesting. I may need you to talk me through this in more depth. But it would represent a design change, so later...